fpmsyncd/fpmsyncd.cpp

```diff
@@ -182,6 +216,67 @@ int main(int argc, char **argv)
                         s.removeSelectable(&eoiuCheckTimer);
                     }
                 }
+                else if (temps == &deviceMetadataTableSubscriber)
+                {
+                    std::deque<KeyOpFieldsValuesTuple> keyOpFvsQueue;
+                    deviceMetadataTableSubscriber.pops(keyOpFvsQueue);
+
+                    for (const auto& keyOpFvs: keyOpFvsQueue)
+                    {
+                        const auto& key = kfvKey(keyOpFvs);
+                        const auto& op = kfvOp(keyOpFvs);
+                        const auto& fvs = kfvFieldsValues(keyOpFvs);
+
+                        if (op != SET_COMMAND)
+                        {
+                            continue;
+                        }
+
+                        if (key != "localhost")
+                        {
+                            continue;
+                        }
+
+                        for (const auto& fv: fvs)
+                        {
+                            const auto& field = fvField(fv);
+                            const auto& value = fvValue(fv);
+
+                            if (field != "suppress-fib-pending")
+                            {
+                                continue;
+                            }
+
+                            bool shouldEnable = (value == "enabled");
+
+                            if (shouldEnable && !sync.isSuppressionEnabled())
+                            {
+                                routeResponseChannel = std::make_unique<NotificationConsumer>(&applStateDb, routeResponseChannelName);
+                                sync.setSuppressionEnabled(true);
+                                s.addSelectable(routeResponseChannel.get());
+                            }
+                            else if (!shouldEnable && sync.isSuppressionEnabled())
+                            {
+                                sync.setSuppressionEnabled(false);
```

Comment:
@stepanblyschak when do the disable, is that possible some in flight messages are lost?

Reply:
Once disable fpmsyncd will reply imidiatelly on receiving route from FRR, no messages are lost

Reply:
@stepanblyschak 
Assume the route A sent to the orchagent to offload, fpmsyncd was waiting for the response. At this time, the route A is in hold status, if we disable the suppress, fpmsyncd won't care the msg sent back to announce route A is offloaded to HW successfully, right? In that case, the route A is offloaded to HW, but the status in FRR is suppressed, right?

Reply:
@StormLiangMS Oh, I get it. Right, we will have suppressed route in FRR in case user disables prior to processing update from orchagent. That can be one or many routes.
We could listen to response channel regardless of suppress status and send back offload flag again but that would generate twice as much traffic to zebra in case feature is disabled.

How about in case of transition from enabled -> disabled fetch all routes present in APPL_DB (it will include those routes that are not set offloaded yet) and send offload status?

Reply:
@stepanblyschak 
The other way we may have a snapshot for the routes in hold in FRR when user disable it, and keep listen to response channel and only send back offload flag which matched with this snapshot list?

In order to avoid the race condition for when we take the snapshot, we may allow a little bit overlap time, during that time frame, send twice traffic to Zebra.

Reply:
> @stepanblyschak The other way we may have a snapshot for the routes in hold in FRR when user disable it, and keep listen to response channel and only send back offload flag which matched with this snapshot list?
> 
> In order to avoid the race condition for when we take the snapshot, we may allow a little bit overlap time, during that time frame, send twice traffic to Zebra.

I rethink it, maybe to make it simple, your idea is good enough, our purpose here it to make sure when suppress enabled, to make sure HW offload happened before advertised out, but if this is disabled, just make sure everything is advertised out would be ok, no need to make sure offload happened before advertise at this time.

Reply:
@stepanblyschak should we revert the sequence, first, setSuppressionEnabled to disable the offload process, then do the markRoutesOffloaded? Otherwise it still timing that some inflight routes after the markRoutesOffloaded.